File: app/src/main/java/com/digicoffer/lauditor/DocEditor/ExpandContentProvider.java
package com.digicoffer.lauditor.DocEditor;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.text.InputFilter;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.widget.AppCompatButton;

import com.digicoffer.lauditor.R;
import com.digicoffer.lauditor.CommonFiles.ValidationUtils.ContentValidation;
import com.digicoffer.lauditor.CommonFiles.ValidationUtils.DescriptionValidation;
import com.google.android.material.textfield.TextInputEditText;

public class ExpandContentProvider extends Dialog {
    private ImageView ivCancelExpand;
    private TextView tvContentTitle, tvContentText;
    private TextInputEditText etContentTitle, etContentText;
    private AppCompatButton btnSaveContent;
    LinearLayout ll_title, ll_content;
    private String titleLabel = "Title", contentLabel = "Content", titleTxt = "";
    private String titleHint = "", contentHint = "";
    private OnSaveClickListener onSaveClickListener;
    LinearLayout linearLayout;
    Boolean isTitleVisible = true;
    Boolean isContentVisible = true;
    InputFilter[] filters, filters1;

    public ExpandContentProvider(Context context, LinearLayout llCreateView, Boolean isTitleVisible, Boolean isContentVisible) {
        super(context, androidx.appcompat.R.style.Base_Theme_AppCompat_Dialog); // You can customize style
        this.linearLayout = llCreateView;
        this.linearLayout.setAlpha(0.5f);
        this.linearLayout.setEnabled(false);
        this.isTitleVisible = isTitleVisible;
        this.isContentVisible = isContentVisible;
    }

    //    @Override
//    protected void onCreate(Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//        setContentView(R.layout.expand_content_layout);
//
//        ivCancelExpand = findViewById(R.id.iv_cancel_expand);
//        tvContentTitle = findViewById(R.id.tv_content_title);
//        tvContentText = findViewById(R.id.tv_content_text);
//        etContentTitle = findViewById(R.id.et_content_title);
//        etContentText = findViewById(R.id.et_content_text);
//        btnSaveContent = findViewById(R.id.btn_save_content);
//
//        tvContentTitle.setText(titleLabel);
//        tvContentText.setText(contentLabel);
//        etContentTitle.setHint(titleHint);
//        etContentText.setHint(contentHint);
//
//        ivCancelExpand.setOnClickListener(v -> dismiss());
//
//        btnSaveContent.setOnClickListener(v -> {
//            if (onSaveClickListener != null) {
//                onSaveClickListener.onSave(
//                        etContentTitle.getText() != null ? etContentTitle.getText().toString().trim() : "",
//                        etContentText.getText() != null ? etContentText.getText().toString().trim() : ""
//                );
//            }
//        });
//        if (getWindow() != null) {
//            getWindow().setLayout(
//                    ViewGroup.LayoutParams.MATCH_PARENT, // width: match parent
//                    ViewGroup.LayoutParams.WRAP_CONTENT  // height: wrap content
//            );
//        }
//    }
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.expand_content_layout);
        setCancelable(false); // prevent back button dismissal
        setCanceledOnTouchOutside(false);
        ivCancelExpand = findViewById(R.id.iv_cancel_expand);
        tvContentTitle = findViewById(R.id.tv_content_title);
        tvContentText = findViewById(R.id.tv_content_text);
        etContentTitle = findViewById(R.id.et_content_title);
        etContentText = findViewById(R.id.et_content_text);
        btnSaveContent = findViewById(R.id.btn_save_content);
        ll_title = findViewById(R.id.ll_title);
        ll_content = findViewById(R.id.ll_content);

        if (!titleTxt.isEmpty()) {
            if (titleTxt.equals("Image")) {
                tvContentTitle.setText("Image Caption");
                tvContentText.setText(titleTxt);
            } else {
                tvContentTitle.setText(titleTxt + " Title");
                tvContentText.setText(titleTxt + " Text");
            }
        }
        if (shouldIncludeTitle(titleLabel)) {
            etContentTitle.setText(titleLabel);
        } else {
            etContentTitle.setText(""); // default label, don't show in edit field
        }
        etContentText.setText(contentLabel);

        etContentTitle.addTextChangedListener(new ContentValidation(etContentTitle,false));
        etContentTitle.setHint(R.string.enter_title);
        etContentText.addTextChangedListener(new DescriptionValidation(etContentText));
        filters = new InputFilter[]{
                new InputFilter.LengthFilter(255)
        };
        filters1 = new InputFilter[]{
                new InputFilter.LengthFilter(5000)
        };
        etContentTitle.setFilters(filters);
        etContentText.setFilters(filters1);
        if (isTitleVisible) {
            ll_title.setVisibility(View.VISIBLE);
        } else {
            ll_title.setVisibility(View.GONE);
        }
        if (isContentVisible) {
            ll_content.setVisibility(View.VISIBLE);
        } else {
            ll_content.setVisibility(View.GONE);
        }
        ivCancelExpand.setOnClickListener(v -> {
            dismiss();
            linearLayout.setAlpha(1.0f);
            linearLayout.setEnabled(true);
        });

        btnSaveContent.setOnClickListener(v -> {
            if (onSaveClickListener != null) {
                onSaveClickListener.onSave(
                        etContentTitle.getText() != null ? etContentTitle.getText().toString().trim() : "",
                        etContentText.getText() != null ? etContentText.getText().toString().trim() : ""
                );
                linearLayout.setAlpha(1.0f);
                linearLayout.setEnabled(true);
            }
        });

        if (getWindow() != null) {
            // 👇 This removes default dialog background
            if (getWindow() != null) {
                getWindow().setBackgroundDrawable(null);

                int marginInDp = 20;
                float scale = getContext().getResources().getDisplayMetrics().density;
                int marginInPx = (int) (marginInDp * scale + 0.5f);

                int screenWidth = getContext().getResources().getDisplayMetrics().widthPixels;
                int dialogWidth = screenWidth - (2 * marginInPx);

                getWindow().setLayout(dialogWidth, ViewGroup.LayoutParams.WRAP_CONTENT);
            }
        }
    }

    private boolean shouldIncludeTitle(String label) {
        switch (label.trim()) {
            case "Overview":
            case "Section":
            case "Sub Section":
            case "Sub Sub Section":
            case "Paragraph":
            case "Numbered List":
            case "Bulleted List":
            case "Page Break":
            case "Image":
            case "Table":
                return false; // don't include title
            default:
                return true;  // custom titles should be shown
        }
    }


    public void setTitles(String titleTxt, String titleLabel, String contentLabel) {
        this.titleTxt = titleTxt;
        this.titleLabel = titleLabel;
        this.contentLabel = contentLabel;
    }

    public void setHints(String titleHint, String contentHint) {
        this.titleHint = titleHint;
        this.contentHint = contentHint;
    }

    public void setOnSaveClickListener(OnSaveClickListener listener) {
        this.onSaveClickListener = listener;
    }

    public interface OnSaveClickListener {
        void onSave(String title, String content);
    }
}

File: app/src/main/java/com/digicoffer/lauditor/DocEditor/ExpandContentProvider.kt
package com.digicoffer.lauditor.DocEditor

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.text.InputFilter
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.widget.AppCompatButton
import com.digicoffer.lauditor.R
import com.digicoffer.lauditor.CommonFiles.ValidationUtils.ContentValidation
import com.digicoffer.lauditor.CommonFiles.ValidationUtils.DescriptionValidation
import com.google.android.material.textfield.TextInputEditText

class ExpandContentProvider(
    context: Context,
    private val linearLayout: LinearLayout,
    private val isTitleVisible: Boolean,
    private val isContentVisible: Boolean
) : Dialog(context, androidx.appcompat.R.style.Base_Theme_AppCompat_Dialog) {

    private var ivCancelExpand: ImageView? = null
    private var tvContentTitle: TextView? = null
    private var tvContentText: TextView? = null
    private var etContentTitle: TextInputEditText? = null
    private var etContentText: TextInputEditText? = null
    private var btnSaveContent: AppCompatButton? = null
    private var ll_title: LinearLayout? = null
    private var ll_content: LinearLayout? = null

    private var titleLabel = "Title"
    private var contentLabel = "Content"
    private var titleTxt = ""
    private var titleHint = ""
    private var contentHint = ""
    private var onSaveClickListener: OnSaveClickListener? = null

    interface OnSaveClickListener {
        fun onSave(title: String, content: String)
    }

    init {
        this.linearLayout.alpha = 0.5f
        this.linearLayout.isEnabled = false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.expand_content_layout)
        setCancelable(false)
        setCanceledOnTouchOutside(false)

        ivCancelExpand = findViewById(R.id.iv_cancel_expand)
        tvContentTitle = findViewById(R.id.tv_content_title)
        tvContentText = findViewById(R.id.tv_content_text)
        etContentTitle = findViewById(R.id.et_content_title)
        etContentText = findViewById(R.id.et_content_text)
        btnSaveContent = findViewById(R.id.btn_save_content)
        ll_title = findViewById(R.id.ll_title)
        ll_content = findViewById(R.id.ll_content)

        if (titleTxt.isNotEmpty()) {
            if (titleTxt == "Image") {
                tvContentTitle?.text = "Image Caption"
                tvContentText?.text = titleTxt
            } else {
                tvContentTitle?.text = "$titleTxt Title"
                tvContentText?.text = "$titleTxt Text"
            }
        }

        if (shouldIncludeTitle(titleLabel)) {
            etContentTitle?.setText(titleLabel)
        } else {
            etContentTitle?.setText("")
        }
        etContentText?.setText(contentLabel)

        etContentTitle?.let { etContentTitle ->
            etContentTitle.addTextChangedListener(ContentValidation(etContentTitle, false))
            etContentTitle.setHint(R.string.enter_title)
            val filters = arrayOf<InputFilter>(InputFilter.LengthFilter(255))
            etContentTitle.filters = filters
        }

        etContentText?.let { etContentText ->
            etContentText.addTextChangedListener(DescriptionValidation(etContentText))
            val filters1 = arrayOf<InputFilter>(InputFilter.LengthFilter(5000))
            etContentText.filters = filters1
        }

        ll_title?.visibility = if (isTitleVisible) View.VISIBLE else View.GONE
        ll_content?.visibility = if (isContentVisible) View.VISIBLE else View.GONE

        ivCancelExpand?.setOnClickListener {
            dismiss()
            linearLayout.alpha = 1.0f
            linearLayout.isEnabled = true
        }

        btnSaveContent?.setOnClickListener {
            if (onSaveClickListener != null) {
                val title = etContentTitle?.text?.toString()?.trim() ?: ""
                val content = etContentText?.text?.toString()?.trim() ?: ""
                onSaveClickListener?.onSave(title, content)
                linearLayout.alpha = 1.0f
                linearLayout.isEnabled = true
            }
        }

        window?.let { window ->
            window.setBackgroundDrawable(null)

            val marginInDp = 20
            val scale = context.resources.displayMetrics.density
            val marginInPx = (marginInDp * scale + 0.5f).toInt()

            val screenWidth = context.resources.displayMetrics.widthPixels
            val dialogWidth = screenWidth - (2 * marginInPx)

            window.setLayout(dialogWidth, ViewGroup.LayoutParams.WRAP_CONTENT)
        }
    }

    private fun shouldIncludeTitle(label: String): Boolean {
        return when (label.trim()) {
            "Overview", "Section", "Sub Section", "Sub Sub Section",
            "Paragraph", "Numbered List", "Bulleted List", "Page Break",
            "Image", "Table" -> false
            else -> true
        }
    }

    fun setTitles(titleTxt: String, titleLabel: String, contentLabel: String) {
        this.titleTxt = titleTxt
        this.titleLabel = titleLabel
        this.contentLabel = contentLabel
    }

    fun setHints(titleHint: String, contentHint: String) {
        this.titleHint = titleHint
        this.contentHint = contentHint
    }

    fun setOnSaveClickListener(listener: OnSaveClickListener?) {
        this.onSaveClickListener = listener
    }
}
